crates/holochain_sqlite/tests/db_wrapper.rs

```diff
@@ -0,0 +1,383 @@
+                    check_task_failed_count.fetch_add(1, Ordering::SeqCst);
+                    tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+                }
+                Err(e) => {
+                    panic!("Got an unexpected error - {:?}", e);
+                }
+                Ok(_c) => {
+                    panic!("Should not have been able to get a connection");
+                }
+            }
+        }
+    });
+
+    futures::future::join_all(vec![readers_task, check_task]).await;
+
+    assert_eq!(3, failed_count.load(Ordering::SeqCst));
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn async_read_releases_permits() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let num_readers = num_read_threads();
+
+    // Run 'read' operations using the connection pool
+    for _ in 0..10 {
+        db_handle
+            .async_reader(move |_| -> Result<(), DatabaseError> {
+                std::thread::sleep(std::time::Duration::from_secs(1));
+                Ok(())
+            })
+            .await
+            .unwrap();
+    }
+
+    // All permits should be available
+    let mut held_permits = Vec::with_capacity(num_readers);
+    for _ in 0..num_readers {
+        let c = db_handle.conn_permit::<DatabaseError>().await.unwrap();
+        held_permits.push(c)
+    }
+
+    assert_eq!(num_readers, held_permits.len());
+}
+
+// TODO this test shows that waiting for a write permit will block until another one is released, whether or not
+//      that is possible.
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn single_write_permit() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let _hold_permit = db_handle.conn_write_permit().await;
+
+    // Bad, will never stop waiting!
+    let result = tokio::time::timeout(
+        std::time::Duration::from_secs(35),
+        db_handle.conn_write_permit(),
+    )
+    .await;
+
+    assert!(result.is_err());
+    // The inner cannot error so this is a timeout error
+    assert!(result.err().is_some())
+}
+
+#[cfg(all(feature = "slow_tests", feature = "test_utils"))]
+#[tokio::test]
+async fn write_permits_can_be_released() {
+    holochain_trace::test_run().unwrap();
+
+    let tmp_dir = tempfile::TempDir::new().unwrap();
+    let db_handle = DbWrite::open(&tmp_dir.into_path(), TestDatabaseKind::new()).unwrap();
+
+    let hold_permit = db_handle.conn_write_permit().await;
+
+    drop(hold_permit);
+
+    // Does not return a result so we can't do much to check the given permit is valid. If the test exits instead of deadlocking then it passed...
+    db_handle.conn_write_permit().await;
```

Comment:
why does this not return a permit the second time?

Reply:
Sorry it does return a permit, just not a `Result` around that permit that we can `Ok` check. In my head I'm thinking about having a timeout inside this function at which point we can check we were successfully get a permit. It's also somewhat meaningless to get a permit back at this point  because you can bypass the permits and get a connection from the fixed size pool so having a permit does not guarantee you can actually get a connection. I will update the comment to more accurately reflect this